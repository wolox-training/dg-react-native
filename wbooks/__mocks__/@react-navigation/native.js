const navigation = jest.genMockFromModule('@react-navigation/native');

module.exports = {
  ...navigation,
  useNavigation: () => ({ navigate: jest.fn() })
};
