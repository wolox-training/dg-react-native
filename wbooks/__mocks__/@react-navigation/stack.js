const stack = {
  createStackNavigator: jest.fn()
};

module.exports = stack;
