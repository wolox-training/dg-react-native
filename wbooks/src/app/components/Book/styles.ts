import { StyleSheet } from 'react-native';
import { COLORS } from '@constants/colors';

export default StyleSheet.create({
  container: {
    backgroundColor: COLORS.white,
    flexDirection: 'row',
    paddingVertical: 15,
    paddingHorizontal: 30,
    marginVertical: 5,
    borderRadius: 5,
    shadowColor: COLORS.black,
    shadowOffset: {
      width: 0,
      height: 1
    },
    shadowOpacity: 0.05,
    shadowRadius: 1,
    elevation: 1
  },
  image: {
    height: 60,
    width: 39
  },
  descriptionContainer: {
    marginLeft: 20
  },
  title: {
    fontSize: 17,
    color: COLORS.gray,
    fontWeight: '600'
  },
  author: {
    fontSize: 15,
    color: COLORS.gray
  }
});
