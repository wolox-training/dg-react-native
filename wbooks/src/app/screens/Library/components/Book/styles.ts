import { StyleSheet } from 'react-native';
import appTheme from '@constants/theme';

export default StyleSheet.create({
  container: {
    backgroundColor: appTheme.colors.white,
    flexDirection: 'row',
    paddingVertical: 15,
    paddingHorizontal: 30,
    borderRadius: 5,
    shadowColor: appTheme.colors.black,
    shadowOffset: {
      width: 0,
      height: 1
    },
    shadowOpacity: 0.05,
    shadowRadius: 1,
    elevation: 1
  },
  image: {
    height: 60,
    width: 39
  },
  descriptionContainer: {
    marginLeft: 20
  },
  title: {
    fontSize: 17,
    fontFamily: 'SFUIText-Bold',
    color: appTheme.colors.gray,
    fontWeight: '400'
  },
  author: {
    fontSize: 15,
    fontFamily: 'SFUIText-Light',
    color: appTheme.colors.gray,
    fontWeight: '400'
  }
});
