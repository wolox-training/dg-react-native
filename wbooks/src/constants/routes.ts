export const ROUTES = {
  library: 'LIBRARY',
  bookDetail: 'BOOK DETAIL'
};
