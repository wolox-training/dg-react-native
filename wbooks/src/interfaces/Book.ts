export interface Book {
  id: number;
  author: string;
  title: string;
  genre: string;
  publisher: string;
  year: string;
  imageUrl?: string | null;
}

export interface FilteredBooks {
  books: Book[];
  query: string;
}

export interface BookState {
  books: Book[];
  filteredBooks: FilteredBooks;
  booksLoading: boolean;
  booksError: string;
}
