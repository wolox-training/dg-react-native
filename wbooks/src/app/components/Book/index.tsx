import React, { useEffect, useState } from 'react';
import { View, TouchableOpacity, Text, Image, Animated } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { validateImageUrl } from '@utils/bookUtils';
import { ROUTES } from '@constants/routes';

import styles from './styles';

interface Props {
  id: number;
  imageUrl?: string | null;
  title: string;
  author: string;
  index: number;
}

const ANIMATION_DELAY = 100;
const ANIMATION_DURATION = 1000;

function Book({ imageUrl, title, author, id, index }: Props) {
  const navigation = useNavigation();
  const source = validateImageUrl(imageUrl);
  const [opacity] = useState(new Animated.Value(0));

  useEffect(() => {
    const fadeIn = () => {
      Animated.timing(opacity, {
        toValue: 1,
        duration: ANIMATION_DURATION,
        delay: ANIMATION_DELAY * index,
        useNativeDriver: true
      }).start();
    };

    fadeIn();
  }, [opacity, index]);

  const onNavigate = () => navigation.navigate(ROUTES.bookDetail, { bookId: id });

  return (
    <Animated.View style={{ opacity }}>
      <TouchableOpacity accessibilityLabel="bookItem" style={styles.container} onPress={onNavigate}>
        <Image accessibilityLabel="bookImage" source={source} style={styles.image} />
        <View style={styles.descriptionContainer}>
          <Text accessibilityLabel="bookTitle" style={styles.title}>
            {title}
          </Text>
          <Text accessibilityLabel="bookAuthor" style={styles.author}>
            {author}
          </Text>
        </View>
      </TouchableOpacity>
    </Animated.View>
  );
}

export default Book;
