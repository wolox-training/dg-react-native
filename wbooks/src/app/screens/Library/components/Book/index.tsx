import React from 'react';
import { View, TouchableOpacity, Text, Image } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { validateImageUrl } from '@utils/bookUtils';
import { ROUTES } from '@constants/routes';

import styles from './styles';

interface Props {
  id: number;
  imageUrl?: string | null;
  title: string;
  author: string;
}

function Book({ imageUrl, title, author }: Props) {
  const navigation = useNavigation();
  const source = validateImageUrl(imageUrl);

  const onNavigate = () => navigation.navigate(ROUTES.bookDetail);

  return (
    <TouchableOpacity style={styles.container} onPress={onNavigate}>
      <Image source={source} style={styles.image} />
      <View style={styles.descriptionContainer}>
        <Text style={styles.title}>{title}</Text>
        <Text style={styles.author}>{author}</Text>
      </View>
    </TouchableOpacity>
  );
}

export default Book;
