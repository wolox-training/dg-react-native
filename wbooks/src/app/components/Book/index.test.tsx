import React from 'react';
import { useNavigation } from '@react-navigation/native';
import { fireEvent, render, waitFor } from '@testing-library/react-native';
import bookPlaceholder from '@assets/book/img_book_placeholder.png';
import { ROUTES } from '@constants/routes';

import Book from './index';

describe('Book test suits', () => {
  const props = {
    author: 'John Doe',
    id: 1,
    title: 'Tibia',
    index: 1
  };

  it('Render correctly', () => {
    const book = render(<Book {...props} />);
    expect(book).toMatchSnapshot();
  });
  it('Render with all props', () => {
    const imageUrl = 'test image';
    const { getByLabelText, getByText } = render(<Book {...props} imageUrl={imageUrl} />);
    const image = getByLabelText('bookImage');
    expect(getByLabelText('bookTitle')).toBeTruthy();
    expect(getByText(props.title)).toBeTruthy();
    expect(getByLabelText('bookAuthor')).toBeTruthy();
    expect(getByText(props.author)).toBeTruthy();
    expect(image).toBeTruthy();
    expect(image.props.source).toEqual({ uri: imageUrl });
  });
  it('Render without imageUrl', () => {
    const { getByLabelText } = render(<Book {...props} />);
    const image = getByLabelText('bookImage');
    expect(image.props.source).toEqual(bookPlaceholder);
  });
  it('Test onNavigate', () => {
    const { getByLabelText } = render(<Book {...props} />);
    const item = getByLabelText('bookItem');
    fireEvent.press(item);
    waitFor(() => {
      expect(useNavigation().navigate).toHaveBeenCalledWith(ROUTES.bookDetail);
    });
  });
});
