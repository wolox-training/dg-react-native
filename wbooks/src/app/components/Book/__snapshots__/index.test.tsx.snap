// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book test suits Render correctly 1`] = `
<View
  style={
    Object {
      "opacity": 0,
    }
  }
>
  <View
    accessibilityLabel="bookItem"
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 5,
        "elevation": 1,
        "flexDirection": "row",
        "marginVertical": 5,
        "opacity": 1,
        "paddingHorizontal": 30,
        "paddingVertical": 15,
        "shadowColor": "#030303",
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.05,
        "shadowRadius": 1,
      }
    }
  >
    <Image
      accessibilityLabel="bookImage"
      source={
        Object {
          "process": [Function],
        }
      }
      style={
        Object {
          "height": 60,
          "width": 39,
        }
      }
    />
    <View
      style={
        Object {
          "marginLeft": 20,
        }
      }
    >
      <Text
        accessibilityLabel="bookTitle"
        style={
          Object {
            "color": "#4A4A4A",
            "fontSize": 17,
            "fontWeight": "600",
          }
        }
      >
        Tibia
      </Text>
      <Text
        accessibilityLabel="bookAuthor"
        style={
          Object {
            "color": "#4A4A4A",
            "fontSize": 15,
          }
        }
      >
        John Doe
      </Text>
    </View>
  </View>
</View>
`;
